File: src/main/java/algorithms/hashtable/146/no_146.py
class LRUCache:
    class Node:
        def __init__(self, key: int = 0, val: int = 0) -> None:
            self.key = key
            self.val = val
            self.prev = None
            self.next = None

    def __init__(self, capacity: int) -> None:
        self.cap = capacity
        self.head = self.Node()
        self.tail = self.Node()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.map = dict()
        self.size = 0

    def __isolate_node(self, cur: 'LRUCache.Node') -> None:
        cur.prev.next = cur.next
        cur.next.prev = cur.prev

    def __move_to_head(self, cur: 'LRUCache.Node') -> None:
        old_first = self.head.next
        self.head.next = cur
        cur.prev = self.head
        old_first.prev = cur
        cur.next = old_first

    def get(self, key: int) -> int:
        if key not in self.map:
            return -1

        cur = self.map[key]
        self.__isolate_node(cur)
        self.__move_to_head(cur)
        return cur.val

    def put(self, key: int, value: int) -> None:
        if self.get(key) != -1:
            self.map[key].val = value
        else:
            if self.size == self.cap:
                # delete the longest unused node
                to_del = self.tail.prev
                self.__isolate_node(to_del)
                self.map.pop(to_del.key)
                self.size -= 1

            ins = self.Node(key, value)
            self.__move_to_head(ins)
            self.map[key] = ins
            self.size += 1

File: src/main/java/algorithms/hashtable/no146/No146.java
package algorithms.hashtable;

import java.util.HashMap;
import java.util.Map;

/**
 * LRU
 * @author Sean Yu
 * @create 2022/3/17 9:41 AM
 */
public class No146 {
    public static void main(String[] args) {
        System.out.println("yodn".compareTo("ewqz"));
//        /**
//         * ["LRUCache","get", "put",  "get",   "put",     "put",   "get"   ,"get"]
//         * [[2]        ,[2]  ,[2,6]  ,[1],      [1,5]    ,[1,2]    ,[1]    ,[2]]
//         */
//        LRUCache lRUCache = new LRUCache(2);
//        System.out.println(lRUCache.get(2));
//        lRUCache.put(2, 6);
//        System.out.println(lRUCache.get(1));
//        lRUCache.put(1, 5);
//        lRUCache.put(1, 2);
//        System.out.println(lRUCache.get(1));
//        System.out.println(lRUCache.get(2));
    }
}

class LRUCache {
    Node head, tail;
    Map<Integer,Node> map;
    int cap;
    public LRUCache(int capacity) {
        map = new HashMap();
        head = new Node();
        tail = new Node();
        head.next = tail;
        tail.prev = head;
        cap = capacity;
    }

    public int get(int key) {
        if(map.containsKey(key)){
            Node node = map.get(key);
            Node p = node.prev;
            Node n = node.next;
            //让node称为孤点，让node前后节点相连
            p.next = n;
            n.prev = p;
            //让node和head相连
            Node nodeNext = head.next;
            head.next = node;
            node.prev = head;
            node.next = nodeNext;
            nodeNext.prev = node;
            return node.val;
        }
        else {
            return -1;
        }
    }

    public void put(int key, int value) {
        int val = get(key);
        //如果node存在，因为get()会自动调整其位置到head之后，所以这里只用调整value
        if(val != -1){
            Node node = map.get(key);
            node.val = value;
        }
        //否则，新建一个并插入到head之后
        else {
            //如果满了，删除最久没用的node
            if(map.size() == cap){
                Node last = tail.prev;
                map.remove(last.key);
                Node p = last.prev;
                p.next = tail;
                tail.prev = p;
            }
            Node node = new Node();
            node.key = key;
            node.val = value;
            map.put(key, node);

            //将node插入head之后
            Node n = head.next;
            head.next = node;
            node.prev = head;
            node.next = n;
            n.prev = node;
        }
    }

    class Node{
        int key;
        int val;
        Node prev;
        Node next;
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
